tests/testsuite/source_replacement.rs

```diff
@@ -0,0 +1,153 @@
+//! Tests for `[source]` table (source replacement).
+
+use std::fs;
+
+use cargo_test_support::registry::{RegistryBuilder, TestRegistry};
+use cargo_test_support::{cargo_process, paths, project, t};
+
+fn setup_replacement(config: &str) -> (TestRegistry, TestRegistry) {
+    let crates_io = RegistryBuilder::new()
+        .no_configure_registry()
+        .http_api()
+        .build();
+
+    let alternative = RegistryBuilder::new().alternative().http_api().build();
+
+    let root = paths::root();
+    t!(fs::create_dir(&root.join(".cargo")));
+    t!(fs::write(root.join(".cargo/config"), config,));
+    (crates_io, alternative)
+}
+
+#[cargo_test]
+fn crates_io_token_not_sent_to_replacement() {
+    // verifies that the crates.io token is not sent to a replacement registry during publish.
+    let (crates_io, _alternative) = setup_replacement(
+        r#"
+        [source.crates-io]
+        replace-with = 'alternative'
+    "#,
+    );
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [project]
```

Comment:
#11135 is trying to unify on using `[package]` within the testsuite.

```suggestion
                [package]
```